### ISSUE TITLE
Virtuoso RDFZZ Error DB.DBA.SPARQL_REXEC

### ISSUE THREAD

**Issue by ghost** [openlink/virtuoso-opensource#752]
I get the following response whenever I try to run a federated query from my sparql endpoint.

>`Virtuoso RDFZZ Error DB.DBA.SPARQL_REXEC('http://data.15926.org/rdl', ...) returned Content-Type 'text/html' status 'HTTP/1.1 200 OK'`

the response and the value I am querying is returned but in a format I am not expecting. I am running VOS version 07.20.3217

**Comment by TallTed** [comment]
It looks like you're not targeting a SPARQL endpoint in your [`DB.DBA.SPARQL_REXEC()`](http://docs.openlinksw.com/virtuoso/fn_sparql_rexec/) call.  I think you just need to replace `'http://data.15926.org/rdl'` with `'http://68.71.136.105/sparql'` (which is the SPARQL endpoint linked from the HTML page at <http://data.15926.org/rdl>).

**Comment by ghost** [comment]
When I used the actual sparql endpoint I don't get the result back from the query only the variable name as a heading. 

Query with sparql endpoint:

```sparql
Select ?Material {
      SERVICE <http://68.71.136.105/sparql> { <http://data.15926.org/rdl/RDS2221635> rdfs:label ?Material}
}
```

Result  =
```
      Material
                     Blank
```
Query without sparql endpoint:

```sparql
Select ?Material {
      SERVICE <http://data.15926.org/rdl> { <http://data.15926.org/rdl/RDS2221635> rdfs:label ?Material}
}
```

Result = 
`Virtuoso RDFZZ Error DB.DBA.SPARQL_REXEC('http://data.15926.org/rdl', ...) returned Content-Type 'text/html' status 'HTTP/1.1 200 OK`

`   316 STAINLESS STEEL`   <=== Data I would like to have bound to `?Material` variable .

SPARQL query:
```sparql
define sql:big-data-const 0 
#output-format:text/html
define sql:signal-void-variables 1 Select ?Material {
      SERVICE <http://data.15926.org/rdl> { <http://data.15926.org/rdl/RDS2221635> rdfs:label ?Material}
}
```
Any thoughts?

**Comment by TallTed** [comment]
You're issuing this query directly against the SPARQL endpoint holding that data, so it doesn't need any `SERVICE` clause, because you're already talking to that `SERVICE`.  [See it here](http://68.71.136.105/sparql?default-graph-uri=&qtxt=SELECT+%3FMaterial+%0D%0AWHERE+%7B+%3Chttp%3A%2F%2Fdata.15926.org%2Frdl%2FRDS2221635%3E+rdfs%3Alabel+%3FMaterial+%7D%0D%0A%0D%0A&format=text%2Fhtml&timeout=0&debug=on).  (In the original question, you were issuing the query against an entirely different service, which does not appear to expect SPARQL queries, nor to receive such from SPARQL clients, and so does not deliver SPARQL results; rather, it expects plain-text queries, and delivers HTML-based results.)

```sparql
SELECT ?Material 
WHERE { <http://data.15926.org/rdl/RDS2221635> rdfs:label ?Material }
```

In other words, your call to [`DB.DBA.SPARQL_REXEC()`](http://docs.openlinksw.com/virtuoso/fn_sparql_rexec/) should look more like this (where I leave out the function arguments you have not shown) -

```
DB.DBA.SPARQL_REXEC('http://data.15926.org/rdl', 'SELECT ?Material WHERE { <http://data.15926.org/rdl/RDS2221635> rdfs:label ?Material }', ...) 
```

**Comment by ghost** [comment]
I think my issue is far simpler than what I thought going through the your response I checked my stored procedures and noticed `DB.DBA.SPARQL_REXEC()` does not exist. Where can I find the create script for this call?

**Comment by TallTed** [comment]
I think you're looking far more deeply than needed.  Perhaps if you start by describing what you're actually trying to achieve?

(Here is the creation of [`DB.DBA.SPARQL_REXEC()`](https://github.com/openlink/virtuoso-opensource/blob/173c3721fcb886f008888ad3bac7877b524d3572/libsrc/Wi/sparql_io.sql#L532), and the creation of the rather more substantial [`DB.DBA.SPARQL_REXEC_INT()`](https://github.com/openlink/virtuoso-opensource/blob/173c3721fcb886f008888ad3bac7877b524d3572/libsrc/Wi/sparql_io.sql#L235) upon which `DB.DBA.SPARQL_REXEC()` depends.)

**Comment by ghost** [comment]
My apologies, I am trying to run a sparql query from my sparql endpoint that queries a remote endpoint to resolve some label names of classes which I used in my triple store. I originally looked at [Sparql Federated query](https://www.w3.org/TR/2013/REC-sparql11-federated-query-20130321/#simpleService) to build my query which is why I was using SERVICE.

**Comment by TallTed** [comment]
OK, so *your* SPARQL endpoint hasn't been referenced above at all.  I'll call it `http://example.com/sparql` going forward.

If I now understand correctly you're trying to run a SPARQL query on that SPARQL endpoint (`http://example.com/sparql`), which will bring results that include data from a remote SPARQL endpoint (`http://68.71.136.105/sparql`; which is not the same service as `http://data.15926.org/rdl`).

You don't need `DB.DBA.SPARQL_REXEC()` at all for this.

See the following [query](http://linkeddata.uriburner.com/sparql?default-graph-uri=&qtxt=SELECT+%3FMaterial+%0D%0A++%7B%0D%0A++++SERVICE+%3Chttp%3A%2F%2F68.71.136.105%2Fsparql%3E%0D%0A++++++%7B+%3Chttp%3A%2F%2Fdata.15926.org%2Frdl%2FRDS2221635%3E+rdfs%3Alabel+%3FMaterial+%7D%0D%0A++%7D&should-sponge=&format=text%2Fhtml&CXML_redir_for_subjs=121&CXML_redir_for_hrefs=&timeout=30000000) (and its [results](http://linkeddata.uriburner.com/sparql?default-graph-uri=&query=SELECT+%3FMaterial+%0D%0A++%7B%0D%0A++++SERVICE+%3Chttp%3A%2F%2F68.71.136.105%2Fsparql%3E%0D%0A++++++%7B+%3Chttp%3A%2F%2Fdata.15926.org%2Frdl%2FRDS2221635%3E+rdfs%3Alabel+%3FMaterial+%7D%0D%0A++%7D&should-sponge=&format=text%2Fhtml&CXML_redir_for_subjs=121&CXML_redir_for_hrefs=&timeout=30000000)), run through a public endpoint which currently permits federated queries --
```sparql
SELECT ?Material 
  {
    SERVICE <http://68.71.136.105/sparql>
      { <http://data.15926.org/rdl/RDS2221635> rdfs:label ?Material }
  }
```

Does that help?

**Comment by ghost** [comment]
Yes, this is exactly what I would like to do. Problem is when I run the exact same query from my sparql endpoint. I only see Material but no value.
![image](https://user-images.githubusercontent.com/10689920/41933821-fb4688b4-7995-11e8-9405-79eb69b325d4.png)

**Comment by TallTed** [comment]
That cannot be the "exact same query" as the column header is "MaterialName".

I think you may have used my query as a sub-query in your overall query -- and that masks the result of my query.  In other words, please try executing my exact query above, on its own.

If this doesn't get you working, is your endpoint accessible over the wire?  Can I run some quick test queries there?

I am wondering about the specific version of Virtuoso you're running, as VOS 07.20.3217 covers a long stretch of time, and may indicate build from `stable/7` or `develop/7`.  Please indicate the branch you built from, and if I cannot reach your endpoint, please provide complete output of the following SPARQL query --
```sparql
SELECT
  ( bif:sys_stat('st_dbms_name')          AS ?name )
  ( bif:sys_stat('st_dbms_ver')           AS ?version )
  ( bif:sys_stat('st_build_date')         AS ?build_date )
  ( bif:sys_stat('git_head')              AS ?git_head )   
  ( bif:sys_stat('st_build_thread_model') AS ?thread )
  ( bif:sys_stat('st_build_opsys_id')     AS ?opsys )
WHERE { ?s ?p ?o } LIMIT 1
```

**Comment by pkleef** [comment]
Please read the following document on [SPARQL-FED capabilities in VOS](http://vos.openlinksw.com/owiki/wiki/VOS/VirtTipsAndTricksDiscoverSPARQFedCapabilitiesSPARQL)

Next, use the `isql` tool to connect to your local Virtuoso instance and perform the following command:

```sql
    SQL> grant SPARQL_SPONGE to "SPARQL";
```
This will add the necessary permissions to the SPARQL endpoint to be able to run SPARQL-FED queries.

Next, you need to run the following command in order for your endpoint to collect some information on the remote SPARQL endpoint:

```sql
    SQL> sparql LOAD SERVICE <http://68.71.136.105/sparql> DATA;
```

That command needs to be performed for each new remote endpoint you want to use, unless you run the following statement which will allow your `/sparql` endpoint to do this automatically in the background:

```sql
    SQL> grant SPARQL_LOAD_SERVICE_DATA to "SPARQL";
```

After running the these commands, your `/sparql` endpoint should be capable of running the following query:

```sparql
SELECT ?Material 
  {
    SERVICE <http://68.71.136.105/sparql>
      { <http://data.15926.org/rdl/RDS2221635> rdfs:label ?Material }
  }
```

During my testing, that returned the following solution:

```
Material
"316 STAINLESS STEEL"@en
```

**Comment by ghost** [comment]
my [endpoint]( http://18.216.183.78:8890/sparql)

Test Result
![image](https://user-images.githubusercontent.com/10689920/41953576-85d4c77e-79e7-11e8-98fd-c9956966bd7a.png)

**Comment by ghost** [comment]
@pkleef I followed the instructions as mentioned and still have the same result. I am no sure what I am doing wrong anymore. Maybe I should try to run a federated sparql query against another endpoint as a test?

This works for me which is interesting:

```
SELECT *
  {
    SERVICE <http://dbpedia.org/sparql>
      { SELECT ?movie ?title 
        WHERE {
                ?movie rdfs:label ?title .
                FILTER LANGMATCHES(LANG(?title), "EN")
              }
      }
  }
```

Result:

![image](https://user-images.githubusercontent.com/10689920/41955607-bea0ae7e-79f1-11e8-82ee-c33aad6904ff.png)

**Comment by TallTed** [comment]
I ran [your latest query against your endpoint](http://18.216.183.78:8890/sparql?default-graph-uri=&qtxt=SELECT+*%0D%0A++%7B%0D%0A++++SERVICE+%3Chttp%3A%2F%2Fdbpedia.org%2Fsparql%3E%0D%0A++++++%7B+SELECT+%3Fmovie+%3Ftitle+%0D%0A++++++++WHERE+%7B%0D%0A++++++++++++++++%3Fmovie+rdfs%3Alabel+%3Ftitle+.%0D%0A++++++++++++++++FILTER+LANGMATCHES%28LANG%28%3Ftitle%29%2C+%22EN%22%29%0D%0A++++++++++++++%7D%0D%0A++++++%7D%0D%0A++%7D%0D%0A&should-sponge=&format=text%2Fhtml&timeout=0&run=+Run+Query+) and got [zero solutions](http://18.216.183.78:8890/sparql?default-graph-uri=&query=SELECT+*%0D%0A++%7B%0D%0A++++SERVICE+%3Chttp%3A%2F%2Fdbpedia.org%2Fsparql%3E%0D%0A++++++%7B+SELECT+%3Fmovie+%3Ftitle+%0D%0A++++++++WHERE+%7B%0D%0A++++++++++++++++%3Fmovie+rdfs%3Alabel+%3Ftitle+.%0D%0A++++++++++++++++FILTER+LANGMATCHES%28LANG%28%3Ftitle%29%2C+%22EN%22%29%0D%0A++++++++++++++%7D%0D%0A++++++%7D%0D%0A++%7D%0D%0A&should-sponge=&format=text%2Fhtml&timeout=0&run=+Run+Query+), while the DBpedia endpoint [gives many solutions](http://dbpedia.org/sparql?default-graph-uri=&query=SELECT+%3Fmovie+%3Ftitle+%0D%0A++++++++WHERE+%7B%0D%0A++++++++++++++++%3Fmovie+rdfs%3Alabel+%3Ftitle+.%0D%0A++++++++++++++++FILTER+LANGMATCHES%28LANG%28%3Ftitle%29%2C+%22EN%22%29%0D%0A++++++++++++++%7D%0D%0A&format=text%2Fhtml&CXML_redir_for_subjs=121&CXML_redir_for_hrefs=&timeout=30000&debug=on&run=+Run+Query+) (most of which aren't movies, because of course, many things have an `rdfs:label` langtagged as English).  Your single-row solution set also doesn't demonstrate success, as you should have received many more rows, as [I did here](http://linkeddata.uriburner.com/sparql?default-graph-uri=&query=SELECT+*%0D%0A++%7B%0D%0A++++SERVICE+%3Chttp%3A%2F%2Fdbpedia.org%2Fsparql%3E%0D%0A++++++%7B+SELECT+%3Fmovie+%3Ftitle+%0D%0A++++++++WHERE+%7B%0D%0A++++++++++++++++%3Fmovie+rdfs%3Alabel+%3Ftitle+.%0D%0A++++++++++++++++FILTER+LANGMATCHES%28LANG%28%3Ftitle%29%2C+%22EN%22%29%0D%0A++++++++++++++%7D%0D%0A++++++%7D%0D%0A++%7D%0D%0A&should-sponge=&format=text%2Fhtml&CXML_redir_for_subjs=121&CXML_redir_for_hrefs=&timeout=30000000) (also see the [query form](http://linkeddata.uriburner.com/sparql?default-graph-uri=&qtxt=SELECT+*%0D%0A++%7B%0D%0A++++SERVICE+%3Chttp%3A%2F%2Fdbpedia.org%2Fsparql%3E%0D%0A++++++%7B+SELECT+%3Fmovie+%3Ftitle+%0D%0A++++++++WHERE+%7B%0D%0A++++++++++++++++%3Fmovie+rdfs%3Alabel+%3Ftitle+.%0D%0A++++++++++++++++FILTER+LANGMATCHES%28LANG%28%3Ftitle%29%2C+%22EN%22%29%0D%0A++++++++++++++%7D%0D%0A++++++%7D%0D%0A++%7D%0D%0A&should-sponge=&format=text%2Fhtml&CXML_redir_for_subjs=121&CXML_redir_for_hrefs=&timeout=30000000)).

I cannot immediately explain why you got the solution you showed, and I got none.  I was thinking perhaps a firewall issue between your endpoint and the remote(s), but that should have produced an error for both of us.

Perhaps @pkleef has more thoughts.

**Comment by ghost** [comment]
I see another user (pasqLisena) who posted on github a [similar Issue](https://github.com/openlink/virtuoso-opensource/issues/232#issuecomment-400720404) seem like there is already an internal ticket made for this, but sure if it is exactly my issue.

**Comment by TallTed** [comment]
@Rshr -

Issue #232 is specifically about Unicode characters in URIs that are part of a SPARQL-FED query, which characters get incorrectly encoded when relayed to the remote `SERVICE`, which encoding is not an element of this issue #752.

Re-reading this issue, it looks like you have provided *part* of each error message you received.  You can help our analysis a lot by including the complete query you're running, along with the full error message you receive (such as seen [here](https://github.com/openlink/virtuoso-opensource/issues/232#issuecomment-223284939)).

(I had misunderstood at first, and thought that *you* were calling the `DB.DBA.SPARQL_REXEC()` function. In fact, this is being called internally by Virtuoso, based on your SPARQL-FED query.)

I see that your Virtuoso was built on May 9, from the `develop/7` branch [as of April 5](https://github.com/openlink/virtuoso-opensource/tree/f46a3f7213848499a06109f86a0a86058b10b194).  I am not immediately aware of any changes made since then that would change your experience, but it would not hurt if you could rebuild from [current code](https://github.com/openlink/virtuoso-opensource/tree/develop/7) and then repeat your test.

It may also be worthwhile to test with a fresh, empty DB -- so we can be absolutely certain of what you're starting with.  Please let us know if you need additional guidance on how to do this.

**Comment by ghost** [comment]
@TallTed I checked out a local version I had setup same version and it works correctly. I get the response as intended, before I rebuild from the current code I would like to know what port if any I need to have open on my web server.

**Comment by TallTed** [comment]
I'm glad to hear that a fresh DB appears to do the right thing.

I am not sure I understand your question about "what port if any ... to have open on my web server".

A typical web server must accept client connections to the web server's port `80` for `http`, and port `443` for `https`, and sometimes other ports for other and/or custom services (i.e., http and https may also be served over other ports).

Are you asking about ports that need to be open to your Virtuoso instance?  At minimum, clients must be able to contact Virtuoso at the `ServerPort` specified in both `[Parameters]` (default `1111`, for iSQL, ODBC, JDBC, OLE DB, ADO.NET, etc.) and the `[HTTPServer]` (default `8890`, for HTTP) stanzas.  Clients may also need to contact the `SSLServerPort` specified in these stanzas, as well as any custom service ports configured or enabled through the Conductor.

Are you asking about ports on other services, that your Virtuoso instance must be able to connect to?  It's generally easiest to let Virtuoso initiate connections targeting any port.  Minimally, for SPARQL-FED queries to work, Virtuoso must be permitted to connect to the remote port(s) in the `SERVICE` -- i.e., `80` for `<http://dbpedia.org/sparql>`.

If you're asking about something else, please be more detailed in your question.

**Comment by ghost** [comment]
@TallTed your response is more than sufficient, I am going setup a fresh DB from the current code. Thanks you for your time and patients.

**Comment by TallTed** [comment]
Happy to help, @Rshr. I'll close this issue now.
